src/Common/src/Interop/Windows/sspicli/SSPIWrapper.cs

```diff
@@ -380,159 +369,145 @@ public static SafeFreeContextBufferChannelBinding QueryContextChannelBinding(SSP
-                    break;
-
-                case Interop.SspiCli.ContextAttribute.SECPKG_ATTR_NEGOTIATION_INFO:
-                    handleType = typeof(SafeFreeContextBuffer);
-                    unsafe
-                    {
-                        nativeBlockSize = sizeof(SecPkgContext_NegotiationInfoW);
-                    }
-                    break;
-
-                case Interop.SspiCli.ContextAttribute.SECPKG_ATTR_CLIENT_SPECIFIED_TARGET:
-                    handleType = typeof(SafeFreeContextBuffer);
-                    break;
-
-                case Interop.SspiCli.ContextAttribute.SECPKG_ATTR_REMOTE_CERT_CONTEXT:
-                    handleType = typeof(SafeFreeCertContext);
-                    break;
+                if (errorCode != 0)
+                {
+                    if (NetEventSource.IsEnabled) NetEventSource.Exit(null, $"ERROR = {ErrorDescription(errorCode)}");
+                    return false;
+                }
 
-                case Interop.SspiCli.ContextAttribute.SECPKG_ATTR_LOCAL_CERT_CONTEXT:
-                    handleType = typeof(SafeFreeCertContext);
-                    break;
+                if (NetEventSource.IsEnabled) NetEventSource.Exit(null, attribute);
+                return true;
+            }
+            finally
+            {
+                sspiHandle?.Dispose();
+            }
+        }
 
-                case Interop.SspiCli.ContextAttribute.SECPKG_ATTR_ISSUER_LIST_EX:
-                    nativeBlockSize = Marshal.SizeOf<Interop.SspiCli.SecPkgContext_IssuerListInfoEx>();
-                    handleType = typeof(SafeFreeContextBuffer);
-                    break;
+        public static string QueryStringContextAttributes(SSPIInterface secModule, SafeDeleteContext securityContext, Interop.SspiCli.ContextAttribute contextAttribute)
+        {
+            Debug.Assert(
+                contextAttribute == Interop.SspiCli.ContextAttribute.SECPKG_ATTR_NAMES ||
+                contextAttribute == Interop.SspiCli.ContextAttribute.SECPKG_ATTR_CLIENT_SPECIFIED_TARGET);
 
-                case Interop.SspiCli.ContextAttribute.SECPKG_ATTR_CONNECTION_INFO:
-                    nativeBlockSize = Marshal.SizeOf<SecPkgContext_ConnectionInfo>();
-                    break;
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(null, contextAttribute);
 
-                case Interop.SspiCli.ContextAttribute.SECPKG_ATTR_APPLICATION_PROTOCOL:
-                    nativeBlockSize = Marshal.SizeOf<Interop.SecPkgContext_ApplicationProtocol>();
-                    break;
+            IntPtr buffer = default;
+            int errorCode = secModule.QueryContextAttributes(
+                securityContext,
+                contextAttribute,
+                MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref buffer, 1)),
+                typeof(SafeFreeContextBuffer),
+                out SafeHandle sspiHandle);
+            try
+            {
+                if (errorCode != 0)
```

Comment:
I think this can be outside of try.. catch (unless there is a chance it returns non-null sspiHandle for error case)

Reply:
> unless there is a chance it returns non-null sspiHandle for error case

That is the majority case.